src/Microsoft.ML.AutoML/TrainerExtensions/SweepableParams.cs

```diff
@@ -105,6 +105,7 @@ public static IEnumerable<SweepableParam> BuildLightGbmParams()
                 new SweepableDiscreteParam("MinimumExampleCountPerLeaf", new object[] { 1, 10, 20, 50 }),
                 new SweepableDiscreteParam("UseCategoricalSplit", new object[] { true, false }),
                 new SweepableDiscreteParam("HandleMissingValue", new object[] { true, false }),
+                new SweepableDiscreteParam("UseZeroAsMissingValue", new object[] { true, false }),
```

Comment:
>UseZeroAsMissingValue [](start = 44, length = 21)

I am not sure that this parameter should be sweepable. Isn't it just to indicate what the file format is?

Reply:
> ... Isn't it just to indicate what the file format is?

Along w/ the original file using zero to indicate missing values, `UseZeroAsMissingValue` would also indicate that we/the user have filled the missing values with `0`, which is the default imputation value in ML․NET.

> I am not sure that this parameter should be sweepable. ...

My thinking is that sweeping this is going to have minimal impact either positively or negatively. 

The main positive impact (to sweeping `UseZeroAsMissingValue`) is when the raw dataset stored its missing values as `0`, the TextLoader mapped missing to zero, or we/the user imputes them to `0` ... and if LightGBM's handling of missing values is more beneficial than treating as an ordinary value (likely, but unknown to me).

The main potential negative impact (to sweeping `UseZeroAsMissingValue`) is that additional sweep options lead to multiplicative combinations for the SMAC sweeper to figure out. The dimensionality of the hyperparameter sweep space increases. The `HandleMissingValue` interacts with `UseZeroAsMissingValue` with `HandleMissingValue` taking precedent and if false, ignores `UseZeroAsMissingValue` ([code](https://github.com/microsoft/LightGBM/blob/c7e903932a28a90ee3477fbd99d6c36219090381/src/io/bin.cpp#L336-L347)); this dependent dimension takes more examples for the sweeper to figure out.

Benchmarks showing `UseZeroAsMissingValue` improving single datasets would be helpful (demonstrating the positive impact). AutoML CLI runs showing the sweeper still gets to the same metric values on a variety of datasets without missing values (demonstrating the potential negative impact) would also be helpful. Though without the CLI currently in master, it's hard to quickly assess this change (@JakeRadMSFT, @eerhardt).

Without benchmarks, my assumption is that it's beneficial to add this sweeping.

---

Missing value handling in LightGBM: ([code](https://github.com/microsoft/LightGBM/blob/c7e903932a28a90ee3477fbd99d6c36219090381/src/io/bin.cpp#L394-L422)), ([code](https://github.com/microsoft/LightGBM/blob/d6c41fa1b287cf1f19fbd694ae6b7423840a014f/src/io/dense_bin.hpp#L151-L229)), ([code](https://github.com/microsoft/LightGBM/blob/d7f8aa53b8be5212887e97e474c90bc5d8beac19/include/LightGBM/tree.h#L221-L241)), ([issue](https://github.com/microsoft/LightGBM/issues/122))

Reply:
Wait, I'm still confused - is this argument used to tell LightGBM how to load data from a file? If this is true, then isn't it meaningless when running in ML.NET?

---
In reply to: [371024932](https://github.com/dotnet/machinelearning/pull/4695#discussion_r371024932) [](ancestors = 371024932)

Reply:
The parameter is used to tell LightGBM how we are representing its missing values in the feature vector we send it. No files involved.

When `UseZeroAsMissingValue == true`, LightGBM will treat feature values of `0.0` as a missing value ([code](https://github.com/microsoft/LightGBM/blob/c7e903932a28a90ee3477fbd99d6c36219090381/src/io/bin.cpp#L336-L347)).

LightGBM then dynamically places missing values in the side of the tree split which maximizes the split gain ([issue](https://github.com/microsoft/LightGBM/issues/122)).

Reply:
I see. So if I understand it correctly now, when running in ML.NET that option would indicate whether the file contains empty strings for missing values, or some other string (like `?` or `nan`) for missing values (since `TextLoader` returns zero when it reads an empty string, and `NaN` when it reads a `?`).

Although, if there are any featurizers in the pipeline, that can "corrupt" what `TextLoader` outputs, can't it? (for instance, change zero to non-zero, or change something to `NaN`?).


---
In reply to: [371927729](https://github.com/dotnet/machinelearning/pull/4695#discussion_r371927729) [](ancestors = 371927729)

Reply:
No files to consider here (or their content). 

Only part to consider is what's passed in the final `Features` column to LightGBM by ML․NET. It's indicating whether we are representing missing values as `0.0` when passed to LightGBM. Otherwise `0.0` is treated as a standard float, like any other feature value.

Generally, users will fill missing values in ML․NET w/ `0.0` since other trainers require imputing, and `0.0` is the `ReplacementMode.DefaultValue` for a float. This is done with:

```c#
pipeline = pipeline.append(mlContext.Transforms.ReplaceMissingValues(new[] {
                new InputOutputColumnPair("FeaturesImputedWithZeros", "FeaturesWithMissingValues")
            },
            MissingValueReplacingEstimator.ReplacementMode.DefaultValue));
```
<!--
Though only the final `Features` column matters, the missing values can originate from the source file, or the featurization process, though it's uncommon (e.g. expression transform w/ `"x : (x > 1 ? ln(x) : na(0.0))"`). Imputed values also come either from the original file or the featurization process (e.g. `ReplaceMissingValues()` above).

I would have to read the LightGBM source more closely, though I think when `UseZeroAsMissingValue` is set, both `0.0` and `Float.NaN` are treated as missing values. Otherwise only `Float.NaN` is treated as a missing value, unless `HandleMissingValue` is set at which point neither have special treatment.
-->